client/packages/common/src/ui/layout/tables/material-react-table/usePaginatedMaterialTable.ts

```diff
@@ -1,98 +1,155 @@
+  columns: PaginatedTableColumnDefinition<T>[];
 }
 
 export const usePaginatedMaterialTable = <T extends MRT_RowData>({
   isLoading,
   onRowClick,
   totalCount,
+  initialSort,
+  columns,
   ...tableOptions
 }: PaginatedTableConfig<T>) => {
   const {
     updateSortQuery,
     updatePaginationQuery,
-    filter,
     queryParams: { sortBy, page, first, offset },
   } = useUrlQueryParams({
-    // TO-DO: Abstract filter/sort logic elsewhere
-    initialSort: { key: 'invoiceNumber', dir: 'desc' },
-    filters: [
-      { key: 'otherPartyName' },
-      { key: 'status', condition: 'equalTo' },
-      { key: 'theirReference' },
-      { key: 'createdDatetime', condition: 'between' },
-      { key: 'shippedDatetime', condition: 'between' },
-      { key: 'invoiceNumber', condition: 'equalTo', isNumber: true },
-    ],
+    initialSort,
   });
-  const { urlQuery, updateQuery } = useUrlQuery();
+  const { updateQuery } = useUrlQuery();
   const [rowSelection, setRowSelection] = useState<MRT_RowSelectionState>({});
+  const paginationRef = useRef<number>(0);
 
   const pagination = { page, first, offset };
 
-  const columnFilters = Object.entries(filter).map(([id, value]) => ({
-    id,
-    value,
-  }));
+  const { mrtColumnDefinitions, filterUpdaters, getFilterState } =
+    useMaterialTableColumns(columns);
 
-  const table = useMaterialReactTable<T>({
-    manualFiltering: true,
-    manualPagination: true,
-    manualSorting: true,
-    // TO-DO: Utility functions for these, and keep elsewhere
-    onSortingChange: sortUpdate => {
+  const handleSortingChange = useCallback(
+    (sortUpdate: MRT_Updater<MRT_SortingState>) => {
       if (typeof sortUpdate === 'function') {
+        // MRT can handle multiple sort fields, but for now we're only
+        // supporting one, so we take the first item of the array
         const newSortValue = sortUpdate([
           { id: sortBy.key, desc: !!sortBy.isDesc },
         ])[0];
-        // console.log('Sorting changed:', newSortValue);
         if (newSortValue)
           updateSortQuery(newSortValue.id, newSortValue.desc ? 'desc' : 'asc');
+        else {
+          // For some reason, when just changing the sort direction on a field,
+          // the sortUpdate method doesn't return anything -- is this a bug in
+          // MRT?
+          updateSortQuery(sortBy.key, !sortBy.isDesc ? 'desc' : 'asc');
+        }
       }
     },
-    onPaginationChange: pagination => {
-      const current = { pageIndex: page, pageSize: first };
-      if (typeof pagination === 'function') {
-        // console.log('current', current);
-        const newPaginationValue = pagination(current);
-        // console.log('Pagination changed:', newPaginationValue);
+    [sortBy, updateSortQuery]
+  );
+
+  const filterState = getFilterState();
+
+  const handlePaginationChange = useCallback(
+    (paginationUpdate: MRT_Updater<MRT_PaginationState>) => {
+      if (typeof paginationUpdate === 'function') {
```

Comment:
I'm finding these a little hard to follow with the comments + dense logic... are these functions considered "done" or still a bit first pass?

Reply:
Yes, they're quite ugly, but I can't really see a way to simplify them. The main problem is that argument  for the "update" functions provided by MRT is a function, which means you have to *call* it to get the current state. Plus I've found a couple of odd behaviours/bugs, which require a bit of a hacky workaround. Let's see if we can't find a way to hide or deal with this a bit better going forward.